blog/2019-10-02-MigratingOptaWebEmployeeRosteringToSpringBoot.adoc

```diff
@@ -0,0 +1,204 @@
+Here is a visual overview of the https://www.optaplanner.org/learn/useCases/employeeRostering.html[Employee Rostering problem]:
+
+image::employeeRosteringProblem.png[]
+
+Many company workplaces need solutions to this problem, from hospitals to call centers.
+There are many factors to consider when assigning shifts to employees, such as required skills, employee availabilities, and employee fairness.
+So how do we find the best solution?
+With OptaPlanner, of course!
+
+== Domain Model
+
+Before we dive into the code, we first need a domain model.
+A domain model represents the entities (Java objects) and how they relate with other entities in the application.
+It tells us the behaviour of the objects and how they change during solving.
+Here's a class diagram of the Employee Rostering model, where employees are nurses:
+
+image::employeeRosteringClassDiagram.png[]
+
+Each of the blocks above represents a Java entity object in our application, which will be saved in a database for solving.
+We use Spring JPA as our persistence layer.
+The arrows show how certain entities map to others, so changing one during the solving process may affect another.
+The `*` and `1` symbols on the arrows represent a many-to-one mapping of the entities in the database.
+
+Here's a simple example of the `Skill` entity:
+
+[source,java]
+----
+@Entity
+@Table(uniqueConstraints = @UniqueConstraint(columnNames = {"tenantId", "name"}))
+public class Skill {
+
+    @NotNull
+    @Size(min = 1, max = 120)
+    @Pattern(regexp = "^(?!\\s).*(?<!\\s)$", message = "Name should not contain any leading or trailing whitespaces")
+    private String name;
+
+    @SuppressWarnings("unused")
+    public Skill() {
+
+    }
+
+    public Skill(Integer tenantId, String name) {
+        super(tenantId);
+        this.name = name;
+    }
+
+    @Override
+    public String toString() {
+        return name;
+    }
+
+    // ************************************************************************
+    // Simple getters and setters
+    // ************************************************************************
+
+    public String getName() {
+        return name;
+    }
+
+    public void setName(String name) {
+        this.name = name;
+    }
+}
+----
+
+This Java object stores the name of the skill and the id of the Tenant (the user, such as a hospital or a call center) it belongs to.
+The `@Entity` annotation represents a table in the JPA database, where each instance of the `Skill` entity has its own row.
+The `@Table` annotation specifies the columns of the table (we'll see this in the next section).
+
+== H2 Database
+
+To test the database, we use the H2 Database Console to run SQL statements and view the saved entities on a user interface.
+Here's an example of executing a SQL statement that will select every `Skill` entity:
+
+image::skillsConsole.png[]
+
+Here we see the autogenerated demo data saved in the H2 database when the application starts up.
```

Comment:
Don't use "we", nor "I". Use "you" if you have to. Best is to use active tense whenever possible and make it shorter.
"Here we see" -> "Here you see ..." or "Notice that ... is saved"